homeassistant/components/tcp/sensor.py

```diff
@@ -49,63 +61,67 @@
 )
 
 
-def setup_platform(hass, config, add_entities, discovery_info=None):
+def setup_platform(
+    hass: HomeAssistant,
+    config: ConfigType,
+    add_entities: AddEntitiesCallback,
+    discovery_info: DiscoveryInfoType | None = None,
+) -> None:
     """Set up the TCP Sensor."""
     add_entities([TcpSensor(hass, config)])
 
 
 class TcpSensor(SensorEntity):
     """Implementation of a TCP socket based sensor."""
 
-    required = ()
-
-    def __init__(self, hass, config):
+    def __init__(self, hass: HomeAssistant, config: ConfigType) -> None:
         """Set all the config values if they exist and get initial state."""
-        value_template = config.get(CONF_VALUE_TEMPLATE)
 
+        value_template: Template | None = config.get(CONF_VALUE_TEMPLATE)
         if value_template is not None:
             value_template.hass = hass
 
         self._hass = hass
-        self._config = {
-            CONF_NAME: config.get(CONF_NAME),
-            CONF_HOST: config.get(CONF_HOST),
-            CONF_PORT: config.get(CONF_PORT),
-            CONF_TIMEOUT: config.get(CONF_TIMEOUT),
-            CONF_PAYLOAD: config.get(CONF_PAYLOAD),
+        self._config: TcpSensorConfig = {
+            CONF_NAME: config[CONF_NAME],
```

Comment:
I'm a bit unsure about this one. Is it possible that `setup_platform` is called with `config` which doesn't conform `PLATFORM_SCHEMA`.

@MartinHjelmare ?

Reply:
validation based on `PLATFORM_SCHEMA` is done before `setup_platform()` is called.
When adding unconform caonfiguration, an error is thrown and platrom not loaded - example

```
sensor:
  - platform: tcp
    foo: IP_ADDRESS
```

result:
`2021-05-21 19:44:02 ERROR (MainThread) [homeassistant.config] Invalid config for [sensor.tcp]: [foo] is an invalid option for [sensor.tcp]. Check: sensor.tcp->foo. (See /workspaces/homeassistant-core/config/configuration.yaml, line 29). Please check the docs at https://www.home-assistant.io/integrations/tcp`